ios/UARCTModule/UARCTMessageView.m

```diff
@@ -0,0 +1,178 @@
+        if (self.onLoadError) {
+            self.onLoadError(@{ UARCTMessageViewMessageKey: messageID,
+                                UARCTMessageViewErrorKey: UARCTMessageViewErrorFailedToFetchMessage,
+                                UARCTMessageViewRetryableKey: @(YES) });
+        }
+    }];
+
+}
+
+- (void)requestMessageBody:(UAInboxMessage *)message {
+    self.message = message;
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:self.message.messageBodyURL];
+    request.timeoutInterval = 60;
+
+    UA_WEAKIFY(self)
+    [[UAirship inboxUser] getUserData:^(UAUserData *userData) {
+        UA_STRONGIFY(self)
+        NSString *auth = [UAUtils userAuthHeaderString:userData];
+        [request setValue:auth forHTTPHeaderField:@"Authorization"];
+        [self.webView loadRequest:request];
+    } queue:dispatch_get_main_queue()];
+}
+
+- (void)webView:(WKWebView *)wv decidePolicyForNavigationResponse:(WKNavigationResponse *)navigationResponse decisionHandler:(void (^)(WKNavigationResponsePolicy))decisionHandler {
+
+    if ([navigationResponse.response isKindOfClass:[NSHTTPURLResponse class]]) {
+        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)navigationResponse.response;
+        NSInteger status = httpResponse.statusCode;
+        if (status >= 400 && status <= 599) {
+            decisionHandler(WKNavigationResponsePolicyCancel);
+            if (status >= 500) {
+                if (self.onLoadError) {
+                    self.onLoadError(@{ UARCTMessageViewMessageKey: self.message.messageID,
+                                        UARCTMessageViewErrorKey: UARCTMessageViewErrorMessageLoadFailed,
+                                        UARCTMessageViewRetryableKey: @(YES) });
+                }
+            } else if (status == 410) {
+                if (self.onLoadError) {
+                    self.onLoadError(@{ UARCTMessageViewMessageKey: self.message.messageID,
+                                        UARCTMessageViewErrorKey: UARCTMessageViewErrorMessageNotAvailable,
+                                        UARCTMessageViewRetryableKey: @(NO) });
+                }
+            } else {
+                if (self.onLoadError) {
+                    self.onLoadError(@{ UARCTMessageViewMessageKey: self.message.messageID,
+                                        UARCTMessageViewErrorKey: UARCTMessageViewErrorMessageLoadFailed,
+                                        UARCTMessageViewRetryableKey: @(NO) });
+                }
+            }
+            return;
+        }
+    }
+
+    decisionHandler(WKNavigationResponsePolicyAllow);
+
+}
+
+- (void)webView:(WKWebView *)wv didFinishNavigation:(WKNavigation *)navigation {
+    if ([UAirship messageCenter].disableMessageLinkPreviewAndCallouts) {
+        [self.webView evaluateJavaScript:@"document.body.style.webkitTouchCallout='none';" completionHandler:nil];
+    }
+
+    // Mark message as read after it has finished loading
```

Comment:
This comment implies to me that `markMessageReadWithCompletionHandler` won't be called until after the message has finished loading, which is a bit confusing given the next `if`. Perhaps just "Mark unread messages as read"?